elastic/kibana
# [Reporting] Add guidance when copying URLs that may have issues

## Summary

Closes https://github.com/elastic/kibana/issues/112122

After https://github.com/elastic/kibana/pull/108553, this PR contributes functionality to detect potential issues that may arise due to the "Copy POST URL" functionality in the share menu.

In discussion with @ThomThomson and @tsullivan we've decided that we want to improve the UX, offering guidance when there is unsaved state and prevent "bad" URLs from being copied in edge cases where the URL may be too long.

Summary of the changes:

* Do not show the Copy POST URL button if there is unsaved state. This also prevents nasty state migration issues in future.
* Do not show the Copy POST URL button if the URL is too long

NB, these changes will be surfaced in:

* Dashboard
* Canvas (shouldn't because Canvas never has "dirty" state)
* Visualize


## Screenshots

### Default OK state for reference

<img width="199" alt="Screenshot 2021-09-27 at 10 50 20" src="https://user-images.githubusercontent.com/8155004/134875960-d67ca163-e163-409d-ad11-cc7f6f170cb2.png">

### Copy URL with unsaved state

<img width="341" alt="Screenshot 2021-10-13 at 17 39 26" src="https://user-images.githubusercontent.com/8155004/137168400-ef2bbe60-f488-4306-857d-bb689e618734.png">

### Copy URL when it is too long with unsaved state - edge case 1

<img width="321" alt="Screenshot 2021-09-27 at 16 25 18" src="https://user-images.githubusercontent.com/8155004/134928163-f63decc4-645a-4fe9-b2af-e604fdcb3f2c.png">

### Copy URL when it is too long, even when state is saved - edge case 2

[UPDATE]

Reaching this state is actually a bug and should not be possible. It means that the code responsible for generating the URL does make use of ways to optimize URL length like providing a saved object reference instead of inserting full values.

<img width="322" alt="Screenshot 2021-09-27 at 16 25 34" src="https://user-images.githubusercontent.com/8155004/134928200-e9a4ca72-1da2-4ca6-8e37-5dc955ee6547.png">


### Additional notes

* In future, scheduled reports will be a first class-citizen. We should consider removing the "Copy POST URL" functionality entirely in that case.

### How to test

1. Start Kibana & ES
2. Load some sample data (from the UI)
3. Open a dashboard
4. Open the share menu (see the normal Copy POST URL functionality under the advanced section)
5. Edit the dashboard
6. Attempt to copy the POST URL (see the warning and "Proceed anyway" functionality)
7. To test the long URL apply the following diff:

<details>

<summary>Diff</summary>

```
diff --git i/x-pack/plugins/reporting/public/share_context_menu/reporting_panel_content.tsx w/x-pack/plugins/reporting/public/share_context_menu/reporting_panel_content.tsx
index d9c258308e0..8c371ed85a7 100644
--- i/x-pack/plugins/reporting/public/share_context_menu/reporting_panel_content.tsx
+++ w/x-pack/plugins/reporting/public/share_context_menu/reporting_panel_content.tsx
@@ -201,7 +201,8 @@ class ReportingPanelContentUi extends Component<Props, State> {
 
   public render() {
     const isUnsaved: boolean = this.isNotSaved() || this.props.isDirty || this.state.isStale;
-    const exceedsMaxLength = this.state.absoluteUrl.length >= CHROMIUM_MAX_URL_LENGTH;
+    // const exceedsMaxLength = this.state.absoluteUrl.length >= CHROMIUM_MAX_URL_LENGTH;
+    const exceedsMaxLength = true;
 
     if (this.props.requiresSavedState && isUnsaved) {
       return (

```

</details>

8. Attempt to copy the POST URL from the share menu
9. See the error text


### Checklist

Delete any items that are not applicable to this PR.


CC @ThomThomson @tsullivan
